crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -2035,3 +2047,45 @@ impl ExecutionComponents {
         self.metrics.clone()
     }
 }
+
+// Cache up to 20000 verified certs. We will need to tune this number in the future - a decent
+// guess to start with is that it should be 10-20 times larger than peak transactions per second,
+// on the assumption that we should see most certs twice within about 10-20 seconds at most: Once via RPC, once via consensus.
+const VERIFIED_CERTIFICATE_CACHE_SIZE: usize = 20000;
+
+pub struct VerifiedCertificateCache {
+    inner: RwLock<LruCache<CertificateDigest, ()>>,
+}
+
+impl Default for VerifiedCertificateCache {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl VerifiedCertificateCache {
+    pub fn new() -> Self {
+        Self {
+            inner: RwLock::new(LruCache::new(
+                NonZeroUsize::new(VERIFIED_CERTIFICATE_CACHE_SIZE).unwrap(),
+            )),
+        }
+    }
+
+    pub fn is_cert_verified(&self, digest: &CertificateDigest) -> bool {
+        let inner = self.inner.read();
+        inner.contains(digest)
+    }
+
+    pub fn cache_cert_verified(&self, digest: CertificateDigest) {
+        let mut inner = self.inner.write();
+        inner.put(digest, ());
```

Comment:
If possible, I would want to add one metric here and in `cache_certs_verified` - check if `inner.put` returns `Some` and increment counter in that case. This would show if we have any sizable percentage of cache misses due to concurrent cert verification (and if yes, this is something we could also address in the future)